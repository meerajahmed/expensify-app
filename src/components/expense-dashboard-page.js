import React from "react";
import ExpenseList from "./expense-list";
import ExpenseListFilter from "./expense-list-filter";


const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;