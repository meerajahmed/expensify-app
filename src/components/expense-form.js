import React from "react";
import moment from "moment";

import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import {  withStyles } from 'react-with-styles';

export default class ExpenseForm extends React.Component {

  state = {
    description: "",
    note: "",
    amount: "",
    createdAt: moment(),
    calenderFocused: false,
    error: false
  };

  constructor(props){
    super(props);

    let defaultState = {
      description: "",
      note: "",
      amount: "",
      createdAt: moment(),
      calenderFocused: false,
      error: false
    };

    if( props.expense ){
      defaultState.description = props.expense.description;
      defaultState.note = props.expense.note;
      defaultState.amount = (props.expense.amount / 100).toString();
      defaultState.createdAt = moment(props.expense.createdAt)
    }
    this.state = defaultState;
  }

  onDescriptionChange = (event) => {
    const description = event.target.value;
    this.setState(() => ({
      description
    }))
  };

  onNoteChange = (event) => {
    const note = event.target.value;
    this.setState(() => ({
      note
    }));
  };

  onAmountChange = (event) => {
    const amount = event.target.value;
    if( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/) ){
      this.setState(() => ({
        amount
      }));
    }
  };

  onDateChange = ( createdAt ) => {
    this.setState({createdAt});
  };

  onFocusChange = ({ focused }) => {
    this.setState({ calenderFocused: focused });
  };

  onSubmit = (event) => {
    event.preventDefault();
    if( !(this.state.description && this.state.amount) ){
      this.setState(() => ({ error: true }));
    } else {
      this.setState(() => ({ error: false }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        { this.state.error && <p>Please provide description and amount!</p> }
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button type="submit">Submit</button>
        </form>
      </div>
    );
  }
}