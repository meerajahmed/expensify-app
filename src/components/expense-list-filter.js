import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate } from "../actions/filters";

const ExpenseListFilter = (props) => (
  <div>
    <input type="text" value={props.filters.text} onChange={(event) => {
      props.dispatch(setTextFilter(event.target.value));
    }} />
    <select value={props.filters.sortBy} onChange={(event)=>{
      if( event.target.value === "amount" ){
        props.dispatch(sortByAmount());
      } else if( event.target.value === "date" ) {
        props.dispatch(sortByDate());
      }
    }}>
      <option value="amount">Amount</option>
      <option value="date">Date</option>
    </select>
  </div>
);

const mapStateToProps = (state) => (
  {
    filters: state.filters
  }
);

export default connect(mapStateToProps)(ExpenseListFilter);
