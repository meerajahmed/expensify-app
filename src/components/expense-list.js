import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./expense-list-item";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {
      props.expenses.length > 0 ? (
        props.expenses.map(( expense ) => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      ) : (
        <p>Empty expense</p>
      )
    }
  </div>
);

/*
const ConnectedExpenseList = connect(
  (state) => (
    {
      expenses: state.expenses
    }
  )
)(ExpenseList);

export default ConnectedExpenseList;*/

const mapStateToProps = (state) => (
  {
    expenses: selectExpenses(state.expenses, state.filters)
  }
);

export default connect(mapStateToProps)(ExpenseList);
