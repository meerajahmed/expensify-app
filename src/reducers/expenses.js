// expense reducer
const expenseReducerDefaultState = [];

export default ( state = expenseReducerDefaultState, action ) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      // concat and return new array
      return [...state, action.expense];

    case "REMOVE_EXPENSE":
      return state.filter((expense) => expense.id !== action.id);

    case "EDIT_EXPENSE":
      return state.map(( expense ) => {
        if( expense.id === action.id ) {
          // merge and return new object
          return {
            ...expense,
            ...action.updates
          }
        } else {
          return expense;
        }
      });

    case "SET_EXPENSES":
      return action.expense;
      
    default:
      return state;
  }
};
