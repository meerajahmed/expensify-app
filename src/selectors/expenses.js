import moment from "moment";

// Get visible expenses
export default ( expenses, { text, sortBy, startDate, endDate} ) => {
  return expenses.filter((expenses) => {
    const createdAtMoment = moment(expenses.createdAt);
    // const startDateMatch = typeof startDate !== "number" || expenses.createdAt >= startDate;
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
    // const endDateMatch = typeof endDate !== "number" || expenses.createdAt <= endDate;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;
    const textMatch = expenses.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    /*
     * Sorting in ascending order
     *  a < b => -1
     *  a = b => 0
     *  a > b => 1
     * */

    // sorting in descending order
    if( sortBy === "date" ){
      return a.createdAt < b.createdAt ? 1 : -1
    } else if( sortBy === "amount") {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};
