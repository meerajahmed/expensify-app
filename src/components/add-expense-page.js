import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./expense-form";
import { addExpense } from "../actions/expenses";

const AddExpensePage = (props) => (
  <div>
    <ExpenseForm onSubmit={(expense) => {
       props.dispatch(addExpense(expense));
       props.history.push('/');
    }} />
  </div>
);

export default connect()(AddExpensePage);