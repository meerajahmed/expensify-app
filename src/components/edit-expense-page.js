import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./expense-form";
import {updateExpense, removeExpense} from "../actions/expenses";

export class EditExpensePage extends React.Component {

  onSubmit = (expense) => {
    this.props.updateExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemove = () => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>
          Remove
        </button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => (
  {
    updateExpense: (id, expense) => {updateExpense(id, expense)},
    removeExpense: (id) => {removeExpense({id})}
  }
);

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);