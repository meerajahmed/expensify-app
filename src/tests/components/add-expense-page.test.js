import React from "react";
import {shallow} from "enzyme";
import {AddExpensePage} from "../../components/add-expense-page";
import expenses from "../fixtures/expenses";

let onSubmit, history, wrapper;

beforeEach(() => {
  onSubmit = jest.fn();
  history = {
    push : jest.fn()
  };
  wrapper = shallow( <AddExpensePage onSubmit={onSubmit} history={history} /> );
});

test("should render add expense page correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle on submit", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")( expenses[1] );
  expect(history.push).toHaveBeenCalledWith("/");
  expect(onSubmit).toHaveBeenCalledWith(expenses[1]);
});