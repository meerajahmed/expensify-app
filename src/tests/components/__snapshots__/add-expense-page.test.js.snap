// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense page correctly 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
